woocommerce/woocommerce
# Blocks E2E: Refactor configs and workflow

Follow-up to https://github.com/woocommerce/woocommerce/pull/46125
Closes: #46331
Related: pdV5qK-nT-p2#possible-solutions

## Proposed Changes

This PR enhances the efficiency of running our Blocks E2E tests both locally and in our CI environment.

First, we have eliminated all `side effects` references as a continuation of #46125, ensuring they are permanently resolved. 🎉 

Regarding CI improvements, we've restructured the Playwright configuration (#46331) to enable all tests to run in a single job rather than a 3D theme-based matrix. This adjustment reduces resource usage by half (10 instead of 20 shards) and accelerates test completion due to better-balanced shards (cc: @tammullen). Additionally, we introduced a `Merge Artifacts` step, consolidating all debug artifacts from separate shards into one zip file for convenience.

For local testing improvements, we've reduced global setup rerun times by reusing existing database snapshots created within the initial run (introduced in #46125). We've also simplified the test execution process, making it unnecessary to identify the specific project (`block-theme`, `classic-theme`, or `block-theme-with-templates`) the test belongs to. Running a single suite is now as straightforward as using `pnpm test:e2e path-to/any-suite.spec.ts`. However, executing all theme-based groups remains an option through `pnpm test:e2e:block-theme/class-theme/block-theme-with-templates` when required.

## Testing Instructions

Ensure running tests via `pnpm test:e2e`, `pnpm test:e2e:block-theme`, `pnpm test:e2e:classic-theme`, `pnpm test:e2e:block-theme-with-templates` works as expected. Ensure all CI tests pass.

⚠️ Some tests might still be flaky. However, that flakiness is out of the scope of this PR, as it is no longer caused by the side effects. Those flaky tests need to be investigated individually.

## Next Steps

- Update [E2E docs](https://github.com/woocommerce/woocommerce/blob/8ee74ede0ebc1957089af08207ae2a1a8250610e/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md).
- Explore uploading DB snapshot to the repo to reduce setup time.
